core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -91,7 +91,7 @@ public TableMetadata refresh() {
     }
     this.version = ver;
     this.currentMetadata = TableMetadataParser.read(this,
-        HadoopInputFile.fromPath(metadataFile, conf));
+        fileIo().newInputFile(metadataFile.toString()));
```

Comment:
If one subclasses `HadoopTableOperations` and returns a `FileIO` instance that isn't a subclass of `HadoopFileIO`, any attempt to cast and call methods could break.

Reply:
I also think it's reasonable for `FileIO` to accept `URI` objects as arguments instead of just `String`s. That gives a stronger opinion about what these paths are as well. Then we call `Path#toUri` and go from there.

Reply:
`Path#toUri` is actually not a good thing to use because it has escaping problems in some cases. The right thing to do is to use strings and paths.

Isn't it possible for `HadoopTableOperations` to update the signature to use `HadoopFileIO`? I thought you could always be more specific, but maybe that's just that you can tell Java when the object is more specific, but not require it of subclasses.

Reply:
You can, but then that requires subclasses of `HadoopTableOperations` to also be returning an instance of `HadoopFileIO` also. I'd very much like to keep everything returning the highest level of the abstraction as possible.

Reply:
In the interest of keeping the API surface area smaller and with higher order abstractions, will keep this as `toString`.